src/metabase/integrations/slack.clj

```diff
@@ -163,34 +173,46 @@
    not-empty
    "Non-empty byte array"))
 
+(s/defn join-channel!
+  "Given a channel ID, calls Slack API `conversations.join` endpoint to join the channel as the Metabase Slack app.
+  This must be done before uploading a file to the channel, if using a Slack app integration."
+  [channel-id :- su/NonBlankString]
+  (POST "conversations.join" {:form-params {:channel channel-id}}))
+
 (s/defn upload-file!
   "Calls Slack API `files.upload` endpoint and returns the URL of the uploaded file."
   [file :- NonEmptyByteArray, filename :- su/NonBlankString, channel-id :- su/NonBlankString]
-  {:pre [(seq (slack-token))]}
-  (let [response (http/post (str slack-api-base-url "/files.upload") {:multipart [{:name "token",    :content (slack-token)}
-                                                                                  {:name "file",     :content file}
-                                                                                  {:name "filename", :content filename}
-                                                                                  {:name "channels", :content channel-id}]
-                                                                      :as        :json})]
-    (if (= 200 (:status response))
-      (u/prog1 (get-in response [:body :file :url_private])
-        (log/debug (trs "Uploaded image") <>))
-      (log/warn (trs "Error uploading file to Slack:") (u/pprint-to-str response)))))
+  {:pre [(slack-configured?)]}
+  (let [request  {:multipart [{:name "file",     :content file}
+                              {:name "filename", :content filename}
+                              {:name "channels", :content channel-id}]}
+        response (try
+                   (POST "files.upload" request)
+                   (catch Throwable e
+                     ;; If file upload fails with a "not_in_channel" error, we join the channel and try again.
+                     ;; This is expected to happen the first time a Slack subscription is sent.
+                     (if (= "not_in_channel" (:error-code (ex-data e)))
+                       (do (join-channel! channel-id)
+                           (POST "files.upload" request))
+                       (throw e))))]
+    (u/prog1 (get-in response [:file :url_private])
+      (log/debug (trs "Uploaded image") <>))))
```

Comment:
Removed the extra status check and error logging here since it's now using the `POST` function which handles it.